### PR TITLE
One Retry/VN per UDP datagram

### LINKED ISSUE #2259: amplification attack using Retry and VN triggered by coalesced Initial packets
An Initial packet can be made pretty small, if I counted correctly the smallest size for a valid Initial is 18 bytes. That means that if I coalesce 67 of these packets, I can fill up one UDP datagram that fulfills the minimum size requirement of 1200 byte.

A naive server implementation might first split the coalesced packet into those 67 QUIC packets, decide that it wants to do address validation and then send a Retry packet for every one of them. Considering that a token will be at least around 40 bytes long (if the server is using authenticated encryption), and that there's UDP and IP overhead, that's a nice amplification factor in itself already, in addition to the fact that the attacker just converted a single large datagram into a lot of small datagrams.
I'm not sure if this is covered by the text yet. We say that

> Prior to validating the client address, servers MUST NOT send more than three times as many bytes as the number of bytes they have received.

However, since the retries are supposed to be stateless, it's not clear how this would work.

The same applies to Version Negotiation packets. With the coalesced packet described above, a single packet could be used to elicit the sending of 67 VN packets. We currently have some text that a server MAY limit the number of VN packets it sends, but unless we come up with a better defense against this attack, I think we'll need stronger language here.

### PR DESCRIPTION
**Pull request by martinthomson**
This keeps things simple.

Closes #2259.

### CODE REVIEW

File: draft-ietf-quic-transport.md [hunk 1]
```diff
@@ -3630,6 +3627,9 @@ The Version Negotiation packet does not include the Packet Number and Length
 fields present in other packets that use the long header form.  Consequently,
 a Version Negotiation packet consumes an entire UDP datagram.
 
+A server MUST NOT send more than one Version Negotiation packet in response to a
+single UDP datagram.
```
Comment (kazuho) [inline draft-ietf-quic-transport.md]: I believe this paragraph is the one you intend to add, and it looks fine to me.

OTOH I see lots of unrelated changes that essentially reverts the recently merged changes. Maybe you need to cherry-pick?

File: draft-ietf-quic-transport.md [hunk 2]
```diff
@@ -3872,7 +3872,9 @@ Destination Connection ID of subsequent packets that it sends.
 
 A server MAY send Retry packets in response to Initial and 0-RTT packets.  A
 server can either discard or buffer 0-RTT packets that it receives.  A server
-can send multiple Retry packets as it receives Initial or 0-RTT packets.
+can send multiple Retry packets as it receives Initial or 0-RTT packets.  A
+server MUST NOT send more than one Retry packet in response to a single UDP
+datagram.
```
Comment (marten-seemann) [inline draft-ietf-quic-transport.md]: And this one.

**Comment by martinthomson** [comment]
I have no idea how that happened.  Forced it.

**Comment by marten-seemann** [comment]
Are we ok with a server sending both a VN packet and a Retry in response to a single UDP datagram?

**Comment by kazuho** [comment]
@marten-seemann Does that happen? IIUC VN is sent only when there's a version mismatch. Retry is sent only when there's _no_ version mismatch.

**Comment by marten-seemann** [comment]
@kazuho: A client could coalesce two packets of different versions.

As I've said on the issue, I still feel uneasy with the fact that a client could coalesce around 70 packets into a single datagram, and I'd prefer more rigid rules to solve this problem.

File: draft-ietf-quic-transport.md (hunk 2 above)
Comment (ianswett) [reply draft-ietf-quic-transport.md]: In what use case are you expecting to send multiple Retrys in response to a UDP datagram?  And in those cases, can we require coalescing the Retrys into a single datagram, and so we could say no more than one UDP datagram in response?

**Comment by marten-seemann** [comment]
Sure, the issue here comes from multi-threaded processing, and reusing packet buffers.

I have one thread (or rather, one go-routine) that reads UDP datagrams from the network, splits the coalesced packets and parses the unencrypted part of the header, and then passes the QUIC packets to their respective connections (which are running in a separate go-routine each). In order to reduce allocations, I reuse the buffers I read my packets into. When handling coalesced packets, this means I have to wait until *all* coalesced parts have been processed before reusing the buffer.
Therefore, I can't just parse the first coalesced packet, pass it on for handling in another threat, and then continue parsing. This would create a race condition, since the handling of the packet might be executed before the parsing is finished, and the packet buffer would be returned already. So I have to split the coalesced packets first, and then handle them.

This wouldn't be a problem at all if a coalesced packet just had a few parts, as it's supposed to be. But since we're maximizing flexibility for no reason, an attacker might send me datagram consisting of around 70 packets, which I would all have to completely parse.

**Comment by kazuho** [comment]
@marten-seemann Thank you for elaborating.

I think that we should be dispatching each QUIC datagram (instead of each QUIC packet) to the respective connections. That's how load balancers are expected to work, and what you are implementing is essentially a load balancer. The benefit of dispatching at the datagram level is that you'd have less things to do on the receiving thread.

> This wouldn't be a problem at all if a coalesced packet just had a few parts, as it's supposed to be. But since we're maximizing flexibility for no reason, an attacker might send me datagram consisting of around 70 packets, which I would all have to completely parse.

I understand your point on flexibility. However, in my view, the flexibility is not only how you coalesce, but the redundancy in all the long header fields. We have deliberately chosen the design because it is clear and easy to analyze (especially in regard to encryption). My preference goes to retaining the clarity by avoiding introducing restrictions.

**Comment by janaiyengar** [comment]
I'm merging this. Please reopen the issue if you think this ought to be discussed further, especially in Tokyo.
